music_assistant/music_providers/ytmusic/ytmusic.py

```diff
@@ -0,0 +1,495 @@
+        """Parse a YT Track response to a Track model object."""
+        track = Track(
+            item_id=track_obj["videoId"], provider=self.type, name=track_obj["title"]
+        )
+        if "artists" in track_obj:
+            track.artists = [
+                await self._parse_artist(artist) for artist in track_obj["artists"]
+            ]
+        if "thumbnails" in track_obj and track_obj["thumbnails"]:
+            track.metadata.images = await self._parse_thumbnails(
+                track_obj["thumbnails"]
+            )
+        if (
+            "album" in track_obj
+            and track_obj["album"]
+            and "id" in track_obj["album"]
+            and track_obj["album"]["id"]
+            and "artists" in track_obj
+        ):
+            album = track_obj["album"]
+            album["artists"] = track_obj["artists"]
+            track.album = await self._parse_album(album, album["id"])
+        if "isExplicit" in track_obj:
+            track.metadata.explicit = track_obj["isExplicit"]
+        if "duration" in track_obj and track_obj["duration"].isdigit():
+            track.duration = track_obj["duration"]
+        available = True
+        if "isAvailable" in track_obj:
+            available = track_obj["isAvailable"]
+        track.add_provider_id(
+            MediaItemProviderId(
+                item_id=str(track_obj["videoId"]),
+                prov_type=self.type,
+                prov_id=self.id,
+                available=available,
+            )
+        )
+        return track
+
+    async def _get_signature_timestamp(self):
+        """Get a signature timestamp required to generate valid stream URLs."""
+        response = await self._get_data(url=YTM_DOMAIN)
+        match = re.search(r'jsUrl"\s*:\s*"([^"]+)"', response)
+        if match is None:
+            raise Exception("Could not identify the URL for base.js player.")
+        url = YTM_DOMAIN + match.group(1)
+        response = await self._get_data(url=url)
+        match = re.search(r"signatureTimestamp[:=](\d+)", response)
+        if match is None:
+            raise Exception("Unable to identify the signatureTimestamp.")
+        return int(match.group(1))
+
+    async def _parse_stream_url(self, stream_format: dict, item_id: str) -> str:
+        """Figure out the stream URL to use based on the YT track object."""
+        cipher_parts = {}
+        for part in stream_format["signatureCipher"].split("&"):
+            key, val = part.split("=", maxsplit=1)
+            cipher_parts[key] = unquote(val)
+        signature = await self._decipher_signature(
+            ciphered_signature=cipher_parts["s"], item_id=item_id
+        )
+        url = cipher_parts["url"] + "&sig=" + signature
+        return url
+
+    @classmethod
+    async def _parse_thumbnails(cls, thumbnails_obj: dict) -> List[MediaItemImage]:
+        """Parse and sort a list of thumbnails from highest quality to lowest quality."""
+        thumbnails_sorted = sorted(
```

Comment:
If the thumbnails are the same there is no need to store them all. In that case only store the highest quality as thumbnail creation is all handled globally

Reply:
They only differ in size, which might be handy for the front-end. Or does the front-end create small version automatically?

Reply:
Done